File: Src/Symbol/Misc/TypeDeclSymbolComponent.cs
using System.Collections.Generic;
using Citron.Collections;
using Pretune;
using Citron.Module;
using Citron.Infra;

namespace Citron.Symbol
{   
    public struct TypeDeclSymbolComponent : ICyclicEqualityComparableStruct<TypeDeclSymbolComponent>
    {        
        Dictionary<DeclSymbolNodeName, ITypeDeclSymbol> typeDict;

        public static TypeDeclSymbolComponent Make()
        {
            var comp = new TypeDeclSymbolComponent();
            comp.typeDict = new Dictionary<DeclSymbolNodeName, ITypeDeclSymbol>();
            return comp;
        }
        
        public IEnumerable<ITypeDeclSymbol> GetEnumerable()
        {
            return typeDict.Values;
        }
        
        public ITypeDeclSymbol? GetType(Name name, int typeParamCount)
        {
            return typeDict.GetValueOrDefault(new DeclSymbolNodeName(name, typeParamCount, default));
        }

        public void AddType(ITypeDeclSymbol typeDecl)
        {
            typeDict.Add(typeDecl.GetNodeName(), typeDecl);
        }

        bool ICyclicEqualityComparableStruct<TypeDeclSymbolComponent>.CyclicEquals(ref TypeDeclSymbolComponent other, ref CyclicEqualityCompareContext context)
        {
            if (!typeDict.CyclicEqualsClassValue(other.typeDict, ref context))
                return false;

            return true;
        }
    }
}

File: src/Symbol/TypeDeclSymbolComponent.cpp
module Citron.DeclSymbols:TypeDeclSymbolComponent;
import <optional>;

using namespace std;

namespace Citron {

//TypeDeclSymbolComponent::TypeDeclSymbolComponent()
//{
//
//}

/*public IEnumerable<ITypeDeclSymbol> GetEnumerable()
{
    return typeDict.Values;
}*/

//TypeDeclSymbol* TypeDeclSymbolComponent::GetType(const Name& name)
//{
//    auto i = typeDict.find(name);
//    if (i == typeDict.end()) return nullptr;
//
//    return &types[i->second];
//}

void TypeDeclSymbolComponent::AddType(TypeDeclSymbol&& typeDecl)
{
    /*size_t index = types.size();

    auto& name = GetName(typeDecl);
    typeDict.insert(name, index);
    types.push_back(std::move(typeDecl));*/
}
//
//bool ICyclicEqualityComparableStruct<TypeDeclSymbolComponent>.CyclicEquals(ref TypeDeclSymbolComponent other, ref CyclicEqualityCompareContext context)
//{
//    if (!typeDict.CyclicEqualsClassValue(other.typeDict, ref context))
//        return false;
//
//    return true;
//}
//
//void ISerializable.DoSerialize(ref SerializeContext context)
//{
//    context.SerializeDictRefKeyRefValue(nameof(typeDict), typeDict);
//}



}